File: assignment-3/javascript/virtualization.test.js
const nock = require('nock');
const fetch = require('node-fetch');

describe('Virtualization', () => {
  test('should test against a mocked payment API', async () => {
    // Mock the payment API
    nock('https://testwithshreyansh.com')
      .post('/api/payment', { amount: 100, currency: 'USD' })
      .reply(200, { status: 'success', transactionId: '12345' });

    // Make a request to the mocked API
    const response = await fetch('https://testwithshreyansh.com/api/payment', {
      method: 'POST',
      body: JSON.stringify({ amount: 100, currency: 'USD' }),
      headers: { 'Content-Type': 'application/json' },
    });

    const paymentResult = await response.json();

    // Validate the response from the mocked API
    expect(paymentResult.status).toBe('success');
    expect(paymentResult.transactionId).toBe('12345');
  });
});

File: assignments/assignment-3/solution/java/src/test/java/com/testwithshreyansh/VirtualizationTest.java
package com.testwithshreyansh;

import com.github.tomakehurst.wiremock.WireMockServer;
import io.restassured.RestAssured;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static org.hamcrest.Matchers.equalTo;

public class VirtualizationTest {

    private WireMockServer wireMockServer;

    @BeforeEach
    public void setup() {
        wireMockServer = new WireMockServer(8089);
        wireMockServer.start();
        setupStub();
    }

    @AfterEach
    public void teardown() {
        wireMockServer.stop();
    }

    public void setupStub() {
        stubFor(post(urlEqualTo("/api/payments"))
            .willReturn(aResponse().withHeader("Content-Type", "application/json")
                .withStatus(201)
                .withBody("{\"status\":\"success\",\"transaction_id\":\"txn_12345\"}")));
    }

    @Test
    public void testMockPaymentApi() {
        RestAssured.baseURI = "http://localhost:8089";
        String paymentData = "{\"amount\":100.00,\"currency\":\"USD\",\"card_number\":\"1234-5678-9012-3456\"}";

        RestAssured.given()
            .contentType("application/json")
            .body(paymentData)
        .when()
            .post("/api/payments")
        .then()
            .statusCode(201)
            .body("status", equalTo("success"))
            .body("transaction_id", equalTo("txn_12345"));
    }
}
